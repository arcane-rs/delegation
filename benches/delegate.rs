use criterion::{criterion_group, criterion_main, Criterion};
use delegation::delegate;

#[delegate]
trait AsStr {
    fn as_str(&self) -> &str;
}

impl AsStr for String {
    fn as_str(&self) -> &str {
        self
    }
}

macro_rules! def_enum {
    ($($variant:ident),+) => {
        #[delegate(derive(AsStr))]
        enum Name {
            $($variant(String)),+
        }
    };
}

macro_rules! def_bench {
    ($($variant:ident),*) => {
        fn bench(bencher: &mut ::criterion::Bencher) {
            $(
                bencher.iter(|| {
                    let name = ::criterion::black_box(
                        Name::$variant(String::from("Test"))
                    );

                    assert_eq!(name.as_str(), "Test");
                });
            )*
        }
    };
}

macro_rules! define {
    ($($variant:ident),*) => {
        def_enum!($($variant),*);
        def_bench!($($variant),*);
    };
}

define!(
    A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y,
    Z
);

fn delegate_benchmark(c: &mut Criterion) {
    c.bench_function("delegate", bench);
}

criterion_group!(benches, delegate_benchmark);
criterion_main!(benches);
