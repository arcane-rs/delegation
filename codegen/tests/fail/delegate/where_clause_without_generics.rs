use delegation::delegate;

#[delegate]
trait Named<N> {
    fn name(&self) -> N;
}

struct User(String);
impl Named<String> for User {
    fn name(&self) -> String {
        self.0.clone()
    }
}

#[delegate(derive(
    Named<String>
    where
        U: Named<String> + 'static;
))]
enum Profile<U> {
    Admin(U),
    User(U),
}

fn main() {
    unreachable!()
}
