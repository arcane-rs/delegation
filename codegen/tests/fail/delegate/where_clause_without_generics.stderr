error[E0412]: cannot find type `U` in this scope
  --> tests/fail/delegate/where_clause_without_generics.rs:20:14
   |
20 | enum Profile<U> {
   |              ^ not found in this scope

error[E0412]: cannot find type `U` in this scope
  --> tests/fail/delegate/where_clause_without_generics.rs:18:9
   |
18 |         U: Named<String> + 'static;
   |         ^ not found in this scope
